### PR TITLE
[EE-IR] Local Function Support

### PR DESCRIPTION
**Pull request by kandersen**
This commit introduces support for calls and references to functions and objects
local to the breakpoint. This commit is supported by corresponding changes in
the compiler.

The commit includes:
  - refined test coverage
  - separation of code-paths for old and new evaluator backends.
  - justification for every remaining muted test included in the suite

Known outstanding issues:
  - referencing local classes from the fragment
  - declarations in the fragment itself
  - result of when-expressions in the fragment is discarded
  - access to protected top level members in `root` package
  - (a handful of tests fail due to disparities in stepping and breakpoint
    behavior between the JVM and the IR backends)

---

Edit: the corresponding compiler changes are [here](https://github.com/JetBrains/kotlin/pull/4773)

### CODE REVIEW

**Comment by kandersen** [comment]
@nikita-nazarov The plug-in side changes are ready for review. Compiler side changes are on their way.

The separation of JVM and IR backend is a little ad-hoc in CodeFragmentCompiler: there's just an interface with a method for every point of variability. Perhaps two separate, although similar, versions of CodeFragmentCompiler would have been cleaner -- it will be easier to throw away the JVM implementation later, as that would just be a few lines changed in KotlinEvaluatorBuilder, whereas currently it should probably include massaging CodeFragmentCompiler to get rid of the indirection via the interface. It's a trade-off, where in the current situation we avoid a lot of code duplication.

File: plugins/kotlin/jvm-debugger/test/testData/evaluation/singleBreakpoint/simpleLocals/localFunctionWithArg.kt
Comment (nikita-nazarov) [inline plugins/kotlin/jvm-debugger/test/testData/evaluation/singleBreakpoint/simpleLocals/localFunctionWithArg.kt:13]: I use the left one! 😄

**Review by nikita-nazarov — CHANGES_REQUESTED** [review]
Thanks for your work! I will merge the pr after compiler changes are merged.

File: plugins/kotlin/jvm-debugger/evaluation/src/org/jetbrains/kotlin/idea/debugger/evaluate/FragmentProjectFileDependencyCollector.kt
Comment (kandersen) [reply plugins/kotlin/jvm-debugger/evaluation/src/org/jetbrains/kotlin/idea/debugger/evaluate/FragmentProjectFileDependencyCollector.kt:54]: Done!

File: plugins/kotlin/jvm-debugger/evaluation/src/org/jetbrains/kotlin/idea/debugger/evaluate/KotlinEvaluatorBuilder.kt
Comment (kandersen) [reply plugins/kotlin/jvm-debugger/evaluation/src/org/jetbrains/kotlin/idea/debugger/evaluate/KotlinEvaluatorBuilder.kt:230]: Thanks! Done.

File: plugins/kotlin/jvm-debugger/evaluation/src/org/jetbrains/kotlin/idea/debugger/evaluate/compilation/IRFragmentCompilerCodegen.kt
Comment (kandersen) [reply plugins/kotlin/jvm-debugger/evaluation/src/org/jetbrains/kotlin/idea/debugger/evaluate/compilation/IRFragmentCompilerCodegen.kt:112]: Good catch!

File: plugins/kotlin/jvm-debugger/evaluation/src/org/jetbrains/kotlin/idea/debugger/evaluate/compilation/IRFragmentCompilerCodegen.kt
Comment (kandersen) [reply plugins/kotlin/jvm-debugger/evaluation/src/org/jetbrains/kotlin/idea/debugger/evaluate/compilation/IRFragmentCompilerCodegen.kt:189]: Done!.

File: plugins/kotlin/jvm-debugger/evaluation/src/org/jetbrains/kotlin/idea/debugger/evaluate/compilation/IRFragmentCompilerCodegen.kt
Comment (kandersen) [reply plugins/kotlin/jvm-debugger/evaluation/src/org/jetbrains/kotlin/idea/debugger/evaluate/compilation/IRFragmentCompilerCodegen.kt:165]: Done!

File: plugins/kotlin/jvm-debugger/evaluation/src/org/jetbrains/kotlin/idea/debugger/evaluate/compilation/CodeFragmentCompiler.kt
Comment (kandersen) [reply plugins/kotlin/jvm-debugger/evaluation/src/org/jetbrains/kotlin/idea/debugger/evaluate/compilation/CodeFragmentCompiler.kt:130]: Split!

File: plugins/kotlin/jvm-debugger/evaluation/src/org/jetbrains/kotlin/idea/debugger/evaluate/compilation/CodeFragmentCompiler.kt
Comment (kandersen) [reply plugins/kotlin/jvm-debugger/evaluation/src/org/jetbrains/kotlin/idea/debugger/evaluate/compilation/CodeFragmentCompiler.kt:84]: Done

**Comment by kandersen** [comment]
Thank you for the review! I have push all, and will keep you in the loop on the compiler side changes :+1:

**Comment by nikita-nazarov** [comment]
https://github.com/JetBrains/intellij-community/commit/0d9b42206001b3ebdecec9d802f5adedfa0b6fec
https://github.com/JetBrains/intellij-community/commit/81e683cd5b6dbf4ba4f2affda75f1a1e90a9c93b

**Pull request closed without merging**
